File: macos/Tests/TranskritoTests/LanguageTests.swift
import Foundation
import Testing
@testable import Transkrito

/// Language classifier, RTL check, settings migration and history compatibility — same vectors as the Windows tests.
struct LanguageTests {
    static let vectors: [String: Any] = CorrectionEngineTests.vectors

    static var languageCases: [(String, String)] {
        ((vectors["language"] as! [String: Any])["cases"] as! [[String: Any]]).map { ($0["text"] as! String, $0["expect"] as! String) }
    }
    static var rtlCases: [(String, Bool)] {
        ((vectors["rtl"] as! [String: Any])["cases"] as! [[String: Any]]).map { ($0["text"] as! String, $0["rtl"] as! Bool) }
    }

    @Test(arguments: LanguageTests.languageCases.map(\.0))
    func textLanguageMatchesSharedVectors(text: String) {
        let expect = Self.languageCases.first { $0.0 == text }!.1
        #expect(LanguageDetector.forHistory(text) == expect)
    }

    @Test(arguments: LanguageTests.rtlCases.map(\.0))
    func rtlDetectionMatchesSharedVectors(text: String) {
        let expect = Self.rtlCases.first { $0.0 == text }!.1
        #expect(LanguageDetector.isRightToLeft(text) == expect)
    }

    @Test func rtlRenderingNeverAltersTheString() throws {
        let arabic = "\u{0623}\u{0646}\u{0627} \u{0623}\u{0633}\u{062A}\u{062E}\u{062F}\u{0645} Claude Code \u{0644}\u{0644}\u{0639}\u{0645}\u{0644}."
        var t = Transcription()
        t.text = arabic
        #expect(t.isRightToLeft)
        #expect(t.text == arabic)
        let data = try JsonFile.encoder.encode(t)
        let back = try JsonFile.decoder.decode(Transcription.self, from: data)
        #expect(back.text == arabic)
    }

    @Test func settingsWithoutLanguageMigrateToAuto() throws {
        let old = """
        { "version": 1, "hotkey": { "key": "D", "modifiers": ["control","option"] }, "model": "en-US",
          "insertAtCursor": false, "hotkeyMode": "toggle", "inputDevice": "abc" }
        """
        let s = try JsonFile.decoder.decode(AppSettings.self, from: Data(old.utf8))
        #expect(s.language == Lang.auto)
        #expect(s.hotkey.key == "D")
        #expect(s.insertAtCursor == false)
        #expect(s.hotkeyMode == "toggle")
        #expect(s.inputDevice == "abc")
        let bogus = try JsonFile.decoder.decode(AppSettings.self, from: Data("{ \"language\": \"fr\" }".utf8))
        #expect(bogus.language == Lang.auto)
        let de = try JsonFile.decoder.decode(AppSettings.self, from: Data("{ \"language\": \"de\" }".utf8))
        #expect(de.language == Lang.de)
    }

    @Test func historyLoadsWithAndWithoutLanguage() throws {
        let json = """
        { "version": 1, "items": [
          { "id": "a", "date": "2026-09-11T05:40:00+02:00", "raw": "x", "text": "Ich muss morgen zur Schule.", "durationSec": 1, "engine": "apple-speech", "model": "de-DE", "corrections": [] },
          { "id": "b", "date": "2026-09-11T05:41:00+02:00", "raw": "y", "text": "hello", "durationSec": 1, "engine": "apple-speech", "model": "en-US", "corrections": [], "language": "en" },
          { "id": "c", "date": "2026-09-11T05:42:00+02:00", "raw": "z", "text": "?", "durationSec": 1, "engine": "nemotron", "model": "m", "corrections": [], "language": "unknown" } ] }
        """
        let file = try JsonFile.decoder.decode(HistoryFile.self, from: Data(json.utf8))
        #expect(file.items.count == 3)
        #expect(file.items[0].language == nil)
        #expect(file.items[0].languageLabel == "")
        #expect(file.items[1].language == "en")
        #expect(file.items[1].languageLabel == "English")
        #expect(file.items[2].languageLabel == "")
        // Old entries are written back without an invented language key.
        let back = try JsonFile.encoder.encode(file)
        let obj = try JSONSerialization.jsonObject(with: back) as! [String: Any]
        let items = obj["items"] as! [[String: Any]]
        #expect(items[0]["language"] == nil)
        #expect(items[1]["language"] as? String == "en")
    }

    @Test func displayNamesAreInTheirOwnScript() {
        #expect(Lang.display(Lang.en) == "English")
        #expect(Lang.display(Lang.de) == "Deutsch")
        #expect(Lang.display(Lang.ar) == "\u{0627}\u{0644}\u{0639}\u{0631}\u{0628}\u{064A}\u{0629}")
        #expect(Lang.display(Lang.unknown) == "")
    }
}

File: windows/Transkrito.Tests/LanguageTests.cs
using System.Text.Json;
using Transkrito.Engine;
using Transkrito.History;
using Transkrito.Settings;
using Transkrito.Storage;
using Xunit;

namespace Transkrito.Tests;

/// <summary>Language classifier, RTL check, settings migration and history compatibility. Vectors from shared/correction-tests.json.</summary>
public class LanguageTests
{
    private static readonly JsonDocument Vectors = JsonDocument.Parse(
        File.ReadAllText(Path.Combine(AppContext.BaseDirectory, "correction-tests.json")));

    public static IEnumerable<object[]> LanguageCases() =>
        Vectors.RootElement.GetProperty("language").GetProperty("cases").EnumerateArray()
            .Select(c => new object[] { c.GetProperty("text").GetString()!, c.GetProperty("expect").GetString()! });

    [Theory]
    [MemberData(nameof(LanguageCases))]
    public void Text_language_matches_shared_vectors(string text, string expect)
    {
        Assert.Equal(expect, LanguageDetector.ForHistory(text));
    }

    public static IEnumerable<object[]> RtlCases() =>
        Vectors.RootElement.GetProperty("rtl").GetProperty("cases").EnumerateArray()
            .Select(c => new object[] { c.GetProperty("text").GetString()!, c.GetProperty("rtl").GetBoolean() });

    [Theory]
    [MemberData(nameof(RtlCases))]
    public void Rtl_detection_matches_shared_vectors(string text, bool rtl)
    {
        Assert.Equal(rtl, LanguageDetector.IsRightToLeft(text));
    }

    [Fact]
    public void Rtl_rendering_never_alters_the_string()
    {
        const string arabic = "أنا أستخدم Claude Code للعمل.";
        var t = new Transcription { Text = arabic };
        Assert.True(t.IsRightToLeft);
        Assert.Equal(arabic, t.Text); // logical order preserved; direction is a rendering property only
        var roundTrip = JsonSerializer.Deserialize<Transcription>(JsonSerializer.Serialize(t, JsonFile.Options), JsonFile.Options)!;
        Assert.Equal(arabic, roundTrip.Text);
    }

    [Fact]
    public void Settings_without_language_migrate_to_auto_without_losing_other_fields()
    {
        const string old = """
        { "version": 1, "hotkey": { "key": "D", "modifiers": ["control","alt"] }, "model": "parakeet-tdt-0.6b-v3",
          "insertAtCursor": false, "hotkeyMode": "toggle", "inputDevice": "abc" }
        """;
        var s = AppSettings.Normalize(JsonSerializer.Deserialize<AppSettings>(old, JsonFile.Options)!);
        Assert.Equal(Lang.Auto, s.Language);
        Assert.Equal("D", s.Hotkey.Key);
        Assert.False(s.InsertAtCursor);
        Assert.Equal("toggle", s.HotkeyMode);
        Assert.Equal("abc", s.InputDevice);
        Assert.Equal("parakeet-tdt-0.6b-v3", s.Model);

        var bogus = AppSettings.Normalize(JsonSerializer.Deserialize<AppSettings>("""{ "language": "fr" }""", JsonFile.Options)!);
        Assert.Equal(Lang.Auto, bogus.Language);
        var explicitDe = AppSettings.Normalize(JsonSerializer.Deserialize<AppSettings>("""{ "language": "de" }""", JsonFile.Options)!);
        Assert.Equal(Lang.De, explicitDe.Language);
    }

    [Fact]
    public void History_loads_with_and_without_language()
    {
        const string json = """
        { "version": 1, "items": [
          { "id": "a", "date": "2026-09-11T05:40:00+02:00", "raw": "x", "text": "Ich muss morgen zur Schule.", "durationSec": 1, "engine": "parakeet", "model": "m", "corrections": [] },
          { "id": "b", "date": "2026-09-11T05:41:00+02:00", "raw": "y", "text": "hello", "durationSec": 1, "engine": "nemotron", "model": "m", "corrections": [], "language": "en" },
          { "id": "c", "date": "2026-09-11T05:42:00+02:00", "raw": "z", "text": "?", "durationSec": 1, "engine": "nemotron", "model": "m", "corrections": [], "language": "unknown" } ] }
        """;
        var file = JsonSerializer.Deserialize<HistoryFile>(json, JsonFile.Options)!;
        Assert.Equal(3, file.Items.Count);
        Assert.Null(file.Items[0].Language);
        Assert.Equal("", file.Items[0].LanguageLabel);
        Assert.Equal("en", file.Items[1].Language);
        Assert.Equal("English", file.Items[1].LanguageLabel);
        Assert.Equal("unknown", file.Items[2].Language);
        Assert.Equal("", file.Items[2].LanguageLabel);

        // Old entries stay untouched when written back: no language key is invented for them.
        var back = JsonSerializer.Serialize(file, JsonFile.Options);
        using var doc = JsonDocument.Parse(back);
        var items = doc.RootElement.GetProperty("items").EnumerateArray().ToList();
        Assert.False(items[0].TryGetProperty("language", out _));
        Assert.Equal("en", items[1].GetProperty("language").GetString());
    }

    [Fact]
    public void Display_names_are_in_their_own_script()
    {
        Assert.Equal("English", Lang.Display(Lang.En));
        Assert.Equal("Deutsch", Lang.Display(Lang.De));
        Assert.Equal("العربية", Lang.Display(Lang.Ar));
        Assert.Equal("", Lang.Display(Lang.Unknown));
    }
}
